src/librustc_interface/passes.rs

```diff
@@ -933,6 +933,10 @@ fn encode_and_write_metadata(
     }).max().unwrap_or(MetadataKind::None);
 
     let metadata = match metadata_kind {
+
+        // Metadata is always required for two-staged compile and link.
```

Comment:
The only necessary data is the data passes to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_utils/codegen_backend/trait.CodegenBackend.html#tymethod.join_codegen_and_link:

* Session
* OutputFilenames
* [CodegenResults](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/struct.CodegenResults.html) (Actually OngoingCodegen is passed, but when joining the codegen, CodegenResults is produced)

DepGraph is not necessary (it is used for storing the location of incremental compilation artifacts, https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_codegen_llvm/lib.rs.html#307 but after joining codegen, the no-link invocation of rustc can do this itself)

Reply:
Yes, metadata is for reconstructing CrateInfo (part of CodegenResult), as it need information about dependent libraries (both rlib and native ones).

Reply:
It should be possible to just derive `Serialize` for `CrateInfo` and put it in a `.rlink` file together with the other necessary data.

Reply:
Good idea, let me look into it.

Reply:
I found `CrateInfo` + "other necessary data" is indeed `CodegenResults`. So I updated my the PR to serialize `CodegenResults` into a `.rlink` file.